contexts/ProposalMarketsContext.tsx

```diff
@@ -0,0 +1,602 @@
+            fetchOpenOrders(wallet.publicKey);
+        }
+    }, [markets, fetchOpenOrders, proposal]);
+
+    useEffect(() => {
+        if (!markets && proposal) {
+            fetchMarketsInfo();
+        }
+    }, [markets, fetchMarketsInfo, proposal]);
+
+    const orderBookObject = useMemo(() => {
+        const getSide = (side: LeafNode[], isBidSide?: boolean) => {
+            if (side.length === 0) {
+                return null;
+            }
+            const parsed = side
+                .map((e) => ({
+                    price: e.key.shrn(64).toNumber(),
+                    size: e.quantity.toNumber(),
+                }))
+                .sort((a, b) => a.price - b.price);
+
+            const sorted = isBidSide
+                ? parsed.sort((a, b) => b.price - a.price)
+                : parsed.sort((a, b) => a.price - b.price);
+
+            const deduped = new Map();
+            sorted.forEach((order) => {
+                if (deduped.get(order.price) === undefined) {
+                    deduped.set(order.price, order.size);
+                } else {
+                    deduped.set(order.price, deduped.get(order.price) + order.size);
+                }
+            });
+
+            const total = parsed.reduce((a, b) => ({
+                price: a.price + b.price,
+                size: a.size + b.size,
+            }));
+            return { parsed, total, deduped };
+        };
+
+        const orderBookSide = (orderBookForSide: LeafNode[], isBidSide?: boolean) => {
+            if (orderBookForSide) {
+                const _orderBookSide = getSide(orderBookForSide, isBidSide);
+                if (_orderBookSide) {
+                    return Array.from(_orderBookSide.deduped?.entries()).map((side) => [
+                        (side[0] / 10_000).toFixed(4),
+                        side[1],
+                    ]);
+                }
+            }
+            if (isBidSide) {
+                return [[0, 0]];
+            }
+            return [[69, 0]];
+        };
+
+        const getToB = (bids: LeafNode[], asks: LeafNode[]) => {
+            const _bids = orderBookSide(bids, true);
+            const _asks = orderBookSide(asks);
+            const tobAsk: number = Number(_asks[0][0]);
+            const tobBid: number = Number(_bids[0][0]);
+            return {
+                topAsk: tobAsk,
+                topBid: tobBid,
+            };
+        };
+
+        const getSpreadString = (bids: LeafNode[], asks: LeafNode[]) => {
+            const { topAsk, topBid } = getToB(bids, asks);
+            const spread: number = topAsk - topBid;
+            const spreadPercent: string = ((spread / topBid) * 100).toFixed(2);
+
+            return spread === topAsk
+                ? '∞ (100.00%)'
+                : `${spread.toFixed(2).toString()} (${spreadPercent}%)`;
+        };
+
+        if (markets) {
+            return {
+                passBidsProcessed: getSide(markets.passBids, true),
```

Comment:
It may be worth reviewing this structure, I threw everything in here, but I'm not sure we use much of it.